File: 프로그래머스/lv2/76502. 괄호 회전하기/괄호 회전하기.py
import copy
def solution(s):
    dic ={')':'(','}':'{',']':'['}
    answer = 0
    for _ in range(len(s)):
        s = s[1:]+s[0]
        tmp = copy.deepcopy(s)
        stack = []
        for x in tmp:
            if x=='(' or x=='[' or x=='{':
                stack.append(x)
            elif stack and dic[x]==stack[-1]:
                stack.pop()
            else:
                stack.append(x)
        if not stack:
            answer += 1
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        int n = s.length();
        Deque<Character> origin = new ArrayDeque<>();
        
        // 1. 문자열 회전을 위해 Stack 으로 만들기
        for(char c : s.toCharArray()){
            origin.addLast(c);
        }
        
        for (int i=0; i<n; i++){
            Deque<Character> stack = new ArrayDeque<>();
            
            // 2. 올바른 괄호 문자열 찾기
            for (char c : origin){
                if (stack.isEmpty() && (c==')' || c=='}' || c==']')) {
                    stack.push('X');
                    break;
                }

                char str = stack.isEmpty() ? 'X' : stack.peek();
                if (c == ')' && str == '(') stack.pop();
                if (c == '}' && str == '{') stack.pop();
                if (c == ']' && str == '[') stack.pop();
                if(c == '(' || c == '{' || c == '[') stack.push(c);
            }
            
            if(stack.isEmpty()) answer ++;

            char firstChar = origin.pollFirst();
            origin.addLast(firstChar);
        }
        return answer;
    }
}